Add custom metadata in the MLmodel file

jdonzallaz opened the pull request:
<!-- 🚨 We recommend pull requests be filed from a non-master branch on a repository fork (e.g. <username>:fix-xxx). 🚨 -->

## Related Issues/PRs

<!--
Please reference any related feature requests, issues, or PRs here. For example, `#123`. To automatically close the referenced issues when this PR is merged, please use a closing keyword (close, fix, or resolve). For example, `Resolve #123`. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue for more information.
-->

<!-- Resolve --> #7235 

## What changes are proposed in this pull request?

Add possibility to add custom metadata to the `MLmodel` file.

Example:
```python
Model.log(
    ...,
    metadata={"metadata_key": "metadata_value"},
)
```

## How is this patch tested?

<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask.
-->
- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

<!--
Please describe how you confirmed the proposed feature/bug-fix/change works here. For example, if you fixed an MLflow client API, you could attach the code that didn't work prior to the fix but works now, or if you added a new feature on MLflow UI, you could attach a video that demonstrates the feature.
-->

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users can now add custom metadata to the `MLmodel` file.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

jdonzallaz committed:
Add custom metadata in the MLmodel file

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

mlflow-automation commented:
<!-- documentation preview -->

### Documentation preview will be available [here](https://output.circle-artifacts.com/output/job/0ed75ca2-8a20-41d2-8183-14eaf07bbe28/artifacts/0/docs/build/html/index.html).

<details>
<summary>Notes</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated on every commit to this PR.
- Job URL: https://circleci.com/gh/mlflow/mlflow/35232
- Updated at: 2022-11-02 15:11:46.288937

</details>

BenWilson2 commented:
This implementation won't work for any of the official flavors in MLflow. As mentioned on the Issue that you filed #7235 in my comments, the API implications for a modification such as this are fairly far-reaching and would require a far more comprehensive implementation and testing regimen than what is shown in this PR. 
Let us get back to you on the issue before you spend any more time working on this PR.

jdonzallaz committed:
Add custom metadata for all flavors

Add metadata to log_model and save_model functions
Add tests for all flavors and serving

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

BenWilson2 commented on mlflow/models/model.py:
```diff
@@ -223,6 +223,10 @@ def add_flavor(self, name, **params):
 
+    def add_metadata(self, metadata):
+        """Add metadata to the model"""
+        self.__dict__.update(metadata)
```
I'm not the biggest fan of this and don't like obfuscated mutation of object members. 
This should be a property so that it can have a descriptor access and should have an accompanying setter to provide assignment. The overriding of `__dict__` key entries can wreak all sorts of havok for users who specify an in-use key in the dictionary that would result in an invalidated and corrupted object that gets saved or logged. 

See the implementation for `signature` for guidance.

jdonzallaz replied:
We wanted to avoid modifying too much code in the current implementation, but you're right, this isn't ideal.

I've updated the code with an actual property.

What I don't like is that now the *metadata* keyword is repeated when accessing the values.
```python
model = mlflow.pyfunc.load_model(model_uri="models:/MyModel/latest")
print(model.metadata.metadata["metadata_key"])
print(model.metadata.get_model_info().metadata["metadata_key"])
# or
model_info = mlflow.models.get_model_info(model_uri="models:/MyModel/latest")
print(model_info.metadata["stateful"])
```

It is not a major problem though.  
It could be avoided by using adding a `__getattr__` method to both `Model` and `ModelInfo` classes:
```python
def __getattr__(self, attr):
    if attr not in self.__dict__:
        return self.metadata[attr]
    return super().__getattr__(attr)
```

Resulting in direct access to the metadata values:
```python
model = mlflow.pyfunc.load_model(model_uri="models:/MyModel/latest")
print(model.metadata.metadata_key)
print(model.metadata.get_model_info().metadata_key)
# or
model_info = mlflow.models.get_model_info(model_uri="models:/MyModel/latest")
print(model_info.metadata_key)
```

Another solution is to change the naming, but I couldn't come up with anything better.

jdonzallaz committed:
Use actual property to store custom metadata

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

jdonzallaz committed:
Change tests for model metadata according to last modifications

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

jdonzallaz committed:
Use Model.load to test mleap metadata logging

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

jdonzallaz committed:
Merge remote-tracking branch 'upstream/master' into add-metadata-to-model

BenWilson2 commented on tests/mleap/test_mleap_model_export.py:
```diff
@@ -218,8 +218,8 @@ def test_model_save_load_with_metadata(spark_model_iris, model_path):
 
-    reloaded_model = mlflow.pyfunc.load_model(model_uri=model_path)
-    assert reloaded_model.metadata.metadata["metadata_key"] == "metadata_value"
+    reloaded_model = Model.load(os.path.join(model_path, "MLmodel"))
```
does this fail to load as a pyfunc?

jdonzallaz replied:
Yes, according to the first note in the [api docs](https://www.mlflow.org/docs/latest/python_api/mlflow.mleap.html), it's not possible to load the MLeap model flavor in Python. That explains why the previous test run failed.

BenWilson2 replied:
Yikes, sorry, been a long week. Yes, loading a JVM object in Python would be... problematic. Thanks for the fix

BenWilson2 reviewed (approved):
LGTM! Thank you for the feature addition. Are you up for a follow-up PR to add a bit to the Docs explaining the existence of this feature? I think a lot of users are going to find this VERY useful and I want to make sure that people are aware that it exists :)

dbczumar reviewed (commented):
@jdonzallaz @BenWilson2 This is looking great (will take a closer look shortly). For Java's model server with MLeap models, can we update the Java representation of the Model object to add the `metadata` field so that MLeap models with metadata can be served in Java? https://github.com/mlflow/mlflow/blob/c076aab7f10125e76168e8c5f9c226a3eca6169b/mlflow/java/scoring/src/main/java/org/mlflow/models/Model.java#L18

jdonzallaz committed:
Add access to metadata in Java's Model

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

jdonzallaz commented:
@BenWilson2 Sure, I can try to explain this feature in the docs with another PR.

@dbczumar I pushed a new change which adds the metadata to the `Model.java` class. I'm not sure it covers 100% of your request, but it gives access to the metadata in Java from the `MLmodel` file. I chose the `JsonNode` type so that about everything can be read, plus it lets the possibility to map to a custom user-defined class corresponding to the metadata.  
I don't have a complete MLeap environment to test everything on my side though.

BenWilson2 commented:
@jdonzallaz Quick update (so that you don't think that we've abandoned this PR) 
We're running these changes for review through a few of the major 1st party tool teams that have tight dependencies on MLflow. Once we're through the notification cycle and review, there might be a few requested changes but we'll likely be ready to proceed with updates and a merge after that. 
Thank you for your patience, and, once again, your contribution and idea!

BenWilson2 commented:
Hi @jdonzallaz we got some review and feedback about this change and the general concern is for the potential confusion about already existing functionality (namely, tags) and this addition. Specifically, when would the metadata option be used vs. when would tags be used? 

Some brief questions:
Is there a limitation to deployment tags or model version tags that prohibits their use for your intended use cases? 

If yes, what is that limitation and what capabilities are not supported with tags?

Would bundling additional configuration files along with the model artifact work if tags do not?

BenWilson2 commented:
From a random conversation today from a user of the Evidently package:
"I would really like the ability to set configurations for a given training run as part of this metadata feature to use for simplifying my deployment monitoring process. It is much more complicated to retrieve this from tags via another REST call. Please try to get the metadata PR merged."

akshaya-a commented:
I'd like to echo the concrete utility of the feature from the AzureML side - we've already discussed the notion of formalizing AzureML Environments into MLflow (think docker + conda as an extension to the current python-only capabilities of mlflow model packaging).

A system or user generating MLmodels can annotate the model with the exact docker image producing it further enhancing the reproducibility than the current OSS definition allows, the concrete example being AutoML or Designer Components that generate MLflow-packaged models. (This isn't a knock at the current format that allows better flexibility than only knowing the docker image)

This information is especially cumbersome, as an above comment notes, to separate from the physical model artifacts into a required metastore that is independent of the model files. Even if systems (like AzureML, ADB) choose to separately store it in our cloud database for other UI/audit flows, consider the "Export My Model" scenario where now you'd need to come up with another serialization format that exports this "dockerimageID tag" along with the model directory anyway (because we wouldn't want to lose this reproducibility information during export right?). Of course, I don't know that this is the right answer, but this PR resonates very nicely with my mental model.

Another thought that comes to mind is the implicit immutability storing it with the artifacts implies versus the currently weak "tagging" notion in MLflow models versus properties/params on Runs. Along with a "let's sign the MLmodel file", adding more metadata to this... metadata-purposebuilt (😄)  seems like a great enhancement to model portability

dbczumar commented:
> I'd like to echo the concrete utility of the feature from the AzureML side - we've already discussed the notion of formalizing AzureML Environments into MLflow (think docker + conda as an extension to the current python-only capabilities of mlflow model packaging).
> 
> A system or user generating MLmodels can annotate the model with the exact docker image producing it further enhancing the reproducibility than the current OSS definition allows, the concrete example being AutoML or Designer Components that generate MLflow-packaged models. (This isn't a knock at the current format that allows better flexibility than only knowing the docker image)
> 
> This information is especially cumbersome, as an above comment notes, to separate from the physical model artifacts into a required metastore that is independent of the model files. Even if systems (like AzureML, ADB) choose to separately store it in our cloud database for other UI/audit flows, consider the "Export My Model" scenario where now you'd need to come up with another serialization format that exports this "dockerimageID tag" along with the model directory anyway (because we wouldn't want to lose this reproducibility information during export right?). Of course, I don't know that this is the right answer, but this PR resonates very nicely with my mental model.
> 
> Another thought that comes to mind is the implicit immutability storing it with the artifacts implies versus the currently weak "tagging" notion in MLflow models versus properties/params on Runs. Along with a "let's sign the MLmodel file", adding more metadata to this... metadata-purposebuilt (😄) seems like a great enhancement to model portability

@akshaya-a What do you think of an API extension that enables users to easily bundle extra files (e.g. environment specs) with their MLflow model? We've added this to `mlflow.pytorch.log_model()`: https://github.com/mlflow/mlflow/pull/3436.

jdonzallaz commented:
Hi @BenWilson2 and @dbczumar,
I still think this metadata implementation is a nice addition, even with existing tags. Other users here brought up some interesting points, and I share their opinions.
To answer your specific questions:

> Is there a limitation to deployment tags or model version tags that prohibits their use for your intended use cases? If yes, what is that limitation and what capabilities are not supported with tags?

I never really used tags before. According to my tests (and by reading the docs), we can indeed add tags to a model, but with a different call (i.e. not directly with `mlflow.{flavor}.log_model`. And I didn't find how to retrieve this value directly when loading a model with `mlflow.pyfunc.load_model`. From what I see, another call is needed. (apparently the user of *Evidently* has the same problem).  
It looks like tags are more related to a *run* while the metadata feature is thought to be closer to the models (think *model registry*).

> Would bundling additional configuration files along with the model artifact work if tags do not?

I feel like it's much cleaner to just add a dict to `log_model`, but yes it should work.

santiagxf commented:
To add to @jdonzallaz points, model tags are mutable properties more related to the model's lifecycle or to capture states of the model that can't be capture otherwise. This metadata implementation may be useful to describe additional properties that the model builder needs to capture and that can't be mutable (changing them would require a new model version). Those property tends to describe the model itself, rather than an attribute of its lifecycle. Hope it makes sense.

BenWilson2 commented:
@santiagxf @jdonzallaz @akshaya-a Thank you again for all of your valuable inputs and your patience. We hope to have a consensus sometime next week!

BenWilson2 commented:
@jdonzallaz Could I ask one more small change from you? Could you add the `@experimental` annotation to the methods within `model.py` and ensure that the docstring entries for them are generated (you can build the docs by following the `make html` command within `CONTRIBUTING.md`) with the experimental status indicator?
Thank you so much!

dbczumar commented:
> @jdonzallaz Could I ask one more small change from you? Could you add the `@experimental` annotation to the methods within `model.py` and ensure that the docstring entries for them are generated (you can build the docs by following the `make html` command within `CONTRIBUTING.md`) with the experimental status indicator? Thank you so much!

@jdonzallaz As a slight addition to this, can we also update the docstrings for the `metadata` parameter to indicate that its experimental and may be removed at any time? After the API has been available for some time with positive feedback, we will mark it stable :)

jdonzallaz commented:
> @jdonzallaz Could I ask one more small change from you? Could you add the `@experimental` annotation to the methods within `model.py` and ensure that the docstring entries for them are generated (you can build the docs by following the `make html` command within `CONTRIBUTING.md`) with the experimental status indicator? Thank you so much!

Hi, it seems there are some problems with the `@experimental` decorator. Here is the generated output:
![image](https://user-images.githubusercontent.com/9055735/207288062-94771652-6115-46af-a193-fc82a455fec4.png)

Using the following code:
```python
class ModelInfo:
    ...

    @property
    @experimental
    def metadata(self) -> Optional[Dict[str, Any]]:
        """
        User defined metadata added to the model
        """
        return self._metadata
```

Removing the `@property` *fix* the output (but that's not what we want). Also, playing with the indentation in the docstring seems to fix it, but that won't work for any decorated element (i.e. classes or functions which are at the first level of indentation VS methods which are the the second level).

```python
# This works
        """
        .. Note::
            Experimental: This {api_type} may change or be removed in a future release without warning.

        User defined metadata added to the model
        """

# This works
        """
    .. Note::
        Experimental: This {api_type} may change or be removed in a future release without warning.

    User defined metadata added to the model
        """

# This doesn't work
        """
        .. Note:: Experimental: This {api_type} may change or be removed in a future release without warning.

        User defined metadata added to the model
        """
```

BenWilson2 commented:
Hi @jdonzallaz For the time being so that you don't have to spend hours testing out different syntax within that decorator's message string, you can directly put the note in the pydoc. I have some work to do in that annotation module anyway and I'll convert your experimental message to the updated decorator formatting. - making note to self right now -

jdonzallaz committed:
Add experimental comment to metadata properties and parameters

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

jdonzallaz commented:
Hi @BenWilson2 & @dbczumar,
According to your message, I pushed a new commit that should address all your requests.
I'm not sure why the CI stage `build_doc_r` fails though. I'm able to build the documentation locally with `make html`.

BenWilson2 commented:
Can you check if the R docs need to be rebuilt and their regenerated files attached to the PR? 
If you do:
```bash
cd docs
make rdocs
```
and if new files are generated, please attach them to the PR.

jdonzallaz commented:
> Can you check if the R docs need to be rebuilt and their regenerated files attached to the PR? If you do:
> 
> ```shell
> cd docs
> make rdocs
> ```
> 
> and if new files are generated, please attach them to the PR.

The command gave not output and no new files were generated. It looks like it's already good.

BenWilson2 commented on mlflow/models/model.py:
```diff
@@ -157,6 +159,16 @@ def mlflow_version(self):
+        .. Note::
+            Experimental: This method may change or be removed in a future release without warning.
+
+        User defined metadata added to the model
```
Can we add an example within the doc string here that shows setting a dictionary of values to a model? For reference, see `models.add_libraries_to_model()`. This same example can be applied (with needed modifications) beneath lines 246 as well.

BenWilson2 commented:
@jdonzallaz I think if you merge master to your branch the test will pass. Once we get an example in the pydoc string for the added attribute, we're good to go to merge! 
Please ping me when it's ready and we'll get this out so people can use it!
Thank you again for your patience and your contribution. We're looking forward to future contributions from you (that will very likely not take this long to get merged ;) )

jdonzallaz committed:
Add examples of metadata usage in docstrings

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

jdonzallaz committed:
Merge remote-tracking branch 'upstream/master' into add-metadata-to-model

jdonzallaz commented:
@BenWilson2 It should be alright now. I added examples where mentioned and merged master.
Thanks for your help in getting this merged. I hope other users will like it too.

jdonzallaz committed:
Fix linting

Signed-off-by: Jonathan Donzallaz <jonathan.donzallaz@gmail.com>

dbczumar reviewed (approved):
LGTM! Fantastic work, @jdonzallaz ! A truly spectacular addition to MLflow! Looking forward to user feedback! :D
